### PR TITLE
Make ExtendedBounds immutable

### LINKED ISSUE #19481: org.elasticsearch.search.aggregations.bucket.DateHistogramIT#testDSTEndTransition fails sporadically
Reproduces locally sometimes, need to run with a high iteration count to get it to trigger.

Example build: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-os-compatibility/os=oraclelinux/790/console

```
> REPRODUCE WITH: gradle :core:integTest -Dtests.seed=82560A64C52E117E -Dtests.class=org.elasticsearch.search.aggregations.bucket.DateHistogramIT -Dtests.method="testDSTEndTransition { seed=[82560A64C52E117E:58332BC3718193E2]}" -Dtests.security.manager=true -Dtests.locale=pl -Dtests.timezone=Asia/Bahrain
FAILURE 0.06s | DateHistogramIT.testDSTEndTransition { seed=[82560A64C52E117E:58332BC3718193E2]} <<< FAILURES!
   > Throwable #1: java.lang.AssertionError: Serialization failed with version [5.0.0-alpha5] bytes should be equal for streamable [org.elasticsearch.search.internal.ShardSearchTransportRequest@77143156]
   > Expected: <org.elasticsearch.common.bytes.PagedBytesReference@87c05c74>
   >      but: was <org.elasticsearch.common.bytes.PagedBytesReference@bf413521>
   >     at __randomizedtesting.SeedInfo.seed([82560A64C52E117E:58332BC3718193E2]:0)
   >     at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
   >     at org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertVersionSerializable(ElasticsearchAssertions.java:667)
   >     at org.elasticsearch.test.transport.AssertingLocalTransport.sendRequest(AssertingLocalTransport.java:93)
   >     at org.elasticsearch.transport.TransportService.sendRequest(TransportService.java:470)
   >     at org.elasticsearch.transport.TransportService.sendRequest(TransportService.java:438)
   >     at org.elasticsearch.search.action.SearchTransportService.sendExecuteQuery(SearchTransportService.java:141)
   >     at org.elasticsearch.action.search.SearchQueryThenFetchAsyncAction.sendExecuteFirstPhase(SearchQueryThenFetchAsyncAction.java:66)
   >     at org.elasticsearch.action.search.AbstractSearchAsyncAction.performFirstPhase(AbstractSearchAsyncAction.java:162)
   >     at org.elasticsearch.action.search.AbstractSearchAsyncAction.start(AbstractSearchAsyncAction.java:143)
   >     at org.elasticsearch.action.search.TransportSearchAction.doExecute(TransportSearchAction.java:96)
   >     at org.elasticsearch.action.search.TransportSearchAction.doExecute(TransportSearchAction.java:45)
   >     at org.elasticsearch.action.support.TransportAction.doExecute(TransportAction.java:158)
   >     at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:182)
   >     at org.elasticsearch.action.ingest.IngestActionFilter.apply(IngestActionFilter.java:80)
   >     at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:180)
   >     at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:153)
   >     at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:87)
   >     at org.elasticsearch.client.node.NodeClient.executeLocally(NodeClient.java:75)
   >     at org.elasticsearch.client.node.NodeClient.doExecute(NodeClient.java:64)
   >     at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:403)
   >     at org.elasticsearch.action.ActionRequestBuilder.execute(ActionRequestBuilder.java:80)
   >     at org.elasticsearch.action.ActionRequestBuilder.execute(ActionRequestBuilder.java:54)
   >     at org.elasticsearch.search.aggregations.bucket.DateHistogramIT.testDSTEndTransition(DateHistogramIT.java:1196)
   >     at java.lang.Thread.run(Thread.java:745)
```

Reproduce with:
```
gradle :core:integTest -Dtests.seed=82560A64C52E117E -Dtests.class=org.elasticsearch.search.aggregations.bucket.DateHistogramIT -Dtests.method="testDSTEndTransition" -Dtests.security.manager=true -Dtests.locale=pl -Dtests.timezone=Asia/Bahrain
```

(May need `-Dtests.iters=50` or similar)

### PR DESCRIPTION
**Pull request by nik9000**
We used to mutate it as part of building the aggregation. That
caused assertVersionSerializable to fail because it assumes that
requests aren't mutated after they are sent.

Closes #19481

### CODE REVIEW

**Comment by nik9000** [comment]
Leaving this marked WIP for now because I haven't finished running the tests.

**Comment by nik9000** [comment]
Tests passing, removed WIP.

**Comment by nik9000** [comment]
@cbuescher, can you review this? Some of your changes revealed the issue and it is in your "time zone" area.

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 1]
```diff
@@ -275,6 +275,13 @@ public long readZLong() throws IOException {
         return BitUtil.zigZagDecode(accumulator | (currentByte << i));
     }
 
+    public Long readOptionalLong() throws IOException {
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: can you add `@Nullable` here?

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java [hunk 2]
```diff
@@ -237,6 +237,15 @@ public void writeZLong(long i) throws IOException {
         writeByte((byte) (value & 0x7F));
     }
 
+    public void writeOptionalLong(Long l) throws IOException {
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamOutput.java]: can you add `@Nullable` here?

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java [hunk 3]
```diff
@@ -20,42 +20,108 @@
...
+public class ExtendedBounds implements ToXContent, Writeable {
     static final ParseField EXTENDED_BOUNDS_FIELD = new ParseField("extended_bounds");
     static final ParseField MIN_FIELD = new ParseField("min");
     static final ParseField MAX_FIELD = new ParseField("max");
 
-    Long min;
-    Long max;
+    public static final ConstructingObjectParser<ExtendedBounds, ParseFieldMatcherSupplier> PARSER = new ConstructingObjectParser<>(
+            "extended_bounds", a -> {
+        Long min = null;
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java]: can you add a sanity check here: `assert a.length == 2`?

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java [hunk 4]
```diff
@@ -20,42 +20,108 @@
...
+     */
     public ExtendedBounds(String minAsStr, String maxAsStr) {
+        this(null, null, minAsStr, maxAsStr);
+    }
+
+    /**
+     * Construct with all possible information.
+     */
+    private ExtendedBounds(Long min, Long max, String minAsStr, String maxAsStr) {
+        this.min = min;
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java]: maybe add assertion here for `(min == null) != (minAsStr == null);` and same for`max`.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java [hunk 5]
```diff
@@ -64,85 +130,45 @@ public ExtendedBounds(String minAsStr, String maxAsStr) {
...
+    /**
+     * Parse the bounds and perform any delayed validation. Returns the result of the parsing.
+     */
+    ExtendedBounds parseAndValidate(String aggName, SearchContext context, DocValueFormat format) {
+        Long min = this.min;
+        Long max = this.max;
         assert format != null;
         if (minAsStr != null) {
-            min = format.parseLong(minAsStr, false, context.nowCallable());
+            min = format.parseLong(minAsStr, false, () -> context.nowInMillis());
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java]: let's save a few lambdas: `context::nowInMillis`

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java [hunk 6]
```diff
@@ -175,6 +201,43 @@ public boolean equals(Object obj) {
         }
         ExtendedBounds other = (ExtendedBounds) obj;
         return Objects.equals(min, other.min)
-                && Objects.equals(min, other.min);
```
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java]: wow, just wow

File: test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java [hunk 7]
```diff
@@ -623,6 +624,14 @@ private static BytesReference serialize(Version version, Streamable streamable)
         return output.bytes();
     }
 
+    private static BytesReference serialize(Version version, Writeable writeable) throws IOException {
```
Comment (ywelsch) [inline test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java]: Is this method used anywhere? `assertVersionSerializable` below operates on `Streamable`.

**Comment by ywelsch** [comment]
@nik9000 left smaller suggestions to add assertion here and there. Feel free to push once addressed. LGTM.

**Comment by nik9000** [comment]
Thanks @ywelsch!

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java (hunk 1 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: Will do.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java (hunk 3 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java]: Sure

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java (hunk 4 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java]: If both are null then there isn't a lower bound. I believe that is allowed.

File: core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java (hunk 5 above)
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/ExtendedBounds.java]: ++

File: test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java (hunk 7 above)
Comment (nik9000) [inline test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java]: It is a leftover. I was using it to hunt for the failure. I'll remove it.

**Pull request merged**